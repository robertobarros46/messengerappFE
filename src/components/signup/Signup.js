import React, { Component } from 'react';
import './Signup.css';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux'
import { createUser } from '../../store/actions/userActions'
import axios from '../../utils/axios/axios'
import { Form, Input, Button } from 'antd';

const FormItem = Form.Item;

export class Signup extends Component {
    constructor(props) {
        super(props);
        this.state = {
            nome: {
                value: ''
            },
            sobrenome: {
                value: ''
            },
            email: {
                value: ''
            },
            password: {
                value: ''
            }
        }
        this.handleInputChange = this.handleInputChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.validateEmailAvailability = this.validateEmailAvailability.bind(this);
        this.isFormInvalid = this.isFormInvalid.bind(this);
    }

    handleInputChange(event, validationFun) {
        const target = event.target;
        const inputName = target.name;        
        const inputValue = target.value;

        this.setState({
            [inputName] : {
                value: inputValue,
                ...validationFun(inputValue)
            }
        });
    }

    handleSubmit(event) {
        event.preventDefault();
        const signupRequest = {
            nome: this.state.nome.value,
            sobrenome: this.state.sobrenome.value,
            email: this.state.email.value,
            password: this.state.password.value,
            role: "USER"
        };
        this.props.createUser(signupRequest);
        this.props.history.push('/login');
    }

    isFormInvalid() {
        return !(this.state.nome.validateStatus === 'success' &&
            this.state.sobrenome.validateStatus === 'success' &&
            this.state.email.validateStatus === 'success' &&
            this.state.password.validateStatus === 'success'
        );
    }

    render() {
        return (
            <div className="signup-container">
                <h1 className="page-title">Sign Up</h1>
                <div className="signup-content">
                    <Form onSubmit={this.handleSubmit} className="signup-form">
                        <FormItem 
                            label="Nome"
                            validateStatus={this.state.nome.validateStatus}
                            help={this.state.nome.errorMsg}>
                            <Input 
                                size="large"
                                name="nome"
                                autoComplete="off"
                                placeholder="Seu primeiro nome"
                                value={this.state.nome.value} 
                                onChange={(event) => this.handleInputChange(event, this.validateName)} />    
                        </FormItem>
                        <FormItem 
                            label="Sobrenome"
                            validateStatus={this.state.sobrenome.validateStatus}
                            help={this.state.sobrenome.errorMsg}>
                            <Input 
                                size="large"
                                name="sobrenome" 
                                autoComplete="off"
                                placeholder="Seu segundo nome"
                                value={this.state.sobrenome.value}
                                onChange={(event) => this.handleInputChange(event, this.validateSobrenome)} />    
                        </FormItem>
                        <FormItem 
                            label="Email"
                            hasFeedback
                            validateStatus={this.state.email.validateStatus}
                            help={this.state.email.errorMsg}>
                            <Input 
                                size="large"
                                name="email" 
                                type="email" 
                                autoComplete="off"
                                placeholder="Seu email"
                                value={this.state.email.value} 
                                onBlur={this.validateEmailAvailability}
                                onChange={(event) => this.handleInputChange(event, this.validateEmail)} />    
                        </FormItem>
                        <FormItem 
                            label="Senha"
                            validateStatus={this.state.password.validateStatus}
                            help={this.state.password.errorMsg}>
                            <Input 
                                size="large"
                                name="password" 
                                type="password"
                                autoComplete="off"
                                placeholder="Um senha entre 6 e 20 caractéres" 
                                value={this.state.password.value} 
                                onChange={(event) => this.handleInputChange(event, this.validatePassword)} />    
                        </FormItem>
                        <FormItem>
                            <Button type="primary" 
                                htmlType="submit" 
                                size="large" 
                                className="signup-form-button"
                                disabled={this.isFormInvalid()}>Sign up</Button>
                            Já é membro? <Link to="/login">Faça login!</Link>
                        </FormItem>
                    </Form>
                </div>
            </div>
        );
    }

    // Validation Functions

    validateName(nome) {
        if(nome.length < 4) {
            return {
                validateStatus: 'error',
                errorMsg: `Nome é muito curto (Mínimo ${4} caractéres necessarios.)`
            }
        } else if (nome.length > 40) {
            return {
                validationStatus: 'error',
                errorMsg: `Nome é muito longo (Máximo ${40} caractéres permitidos.)`
            }
        } else {
            return {
                validateStatus: 'success',
                errorMsg: null,
              };            
        }
    }

    validateEmail (email) {
        if(!email) {
            return {
                validateStatus: 'error',
                errorMsg: 'Email não pode estar vazio'                
            }
        }

        const EMAIL_REGEX = RegExp('[^@ ]+@[^@ ]+\\.[^@ ]+');
        if(!EMAIL_REGEX.test(email)) {
            return {
                validateStatus: 'error',
                errorMsg: 'Email não é válido'
            }
        }

        if(email.length > 40) {
            return {
                validateStatus: 'error',
                errorMsg: `Email é muito longo (Máximo ${40} caractéres permitidos)`
            }
        }

        return {
            validateStatus: null,
            errorMsg: null
        }
    }

    validateSobrenome (sobrenome) {
        if(sobrenome.length < 3) {
            return {
                validateStatus: 'error',
                errorMsg: `Sobrenome é muito curto (Mínimo ${3} caractéres necessarios.)`
            }
        } else if (sobrenome.length > 20) {
            return {
                validationStatus: 'error',
                errorMsg: `Sobrenome é muito longo (Máximo ${20} caractéres permitidos.)`
            }
        } else {
            return {
                validateStatus: 'success',
                errorMsg: null
            }
        }
    }

    validateEmailAvailability() {
        // First check for client side errors in email
        const emailValue = this.state.email.value;
        const emailValidation = this.validateEmail(emailValue);

        if(emailValidation.validateStatus === 'error') {
            this.setState({
                email: {
                    value: emailValue,
                    ...emailValidation
                }
            });    
            return;
        }

        this.setState({
            email: {
                value: emailValue,
                validateStatus: 'validating',
                errorMsg: null
            }
        });

        axios({
            method: 'get',
            url: "/api/v1/users/emailavailability?email=" + emailValue
        })
        .then(response => {
            if(response.data.available) {
                this.setState({
                    email: {
                        value: emailValue,
                        validateStatus: 'success',
                        errorMsg: null
                    }
                });
            } else {
                this.setState({
                    email: {
                        value: emailValue,
                        validateStatus: 'error',
                        errorMsg: 'Email indisponível'
                    }
                });
            }
        }).catch(error => {
            // Marking validateStatus as success, Form will be recchecked at server
            this.setState({
                email: {
                    value: emailValue,
                    validateStatus: 'success',
                    errorMsg: error.message
                }
            });
        });
    }

    validatePassword (password) {
        if(password.length < 6) {
            return {
                validateStatus: 'error',
                errorMsg: `Senha é muito curta (Mínimo ${6} caractéres necessarios.)`
            }
        } else if (password.length > 20) {
            return {
                validationStatus: 'error',
                errorMsg: `Senha é muito longa (Máximo ${20} caractéres permitidos.)`
            }
        } else {
            return {
                validateStatus: 'success',
                errorMsg: null,
            };            
        }
    }

}

const mapDispatchToProps = (dispatch) => {
    return {
        createUser: (user) => dispatch(createUser(user))
    }
}

export default connect(null, mapDispatchToProps)(Signup);