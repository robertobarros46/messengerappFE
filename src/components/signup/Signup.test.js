import React from 'react';
import { shallow, configure } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import { Signup } from './Signup'
configure({ adapter: new Adapter() });

describe('<Signup />', () => {
    let wrapper;
    const history = {
        push: () => {}
    }
    beforeEach(() => {
        wrapper = shallow(<Signup createUser={() => {}} history={history}/>);
    });

    it('+++ render LoginPage correctly, snapshot', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    // Check render component without errors
    it('+++ render the DUMB component', () => {
        expect(wrapper.length).toEqual(1);
    });

    it('+++ Form should have 5 children and view should have 1 Header, 4 input fields and 1 Button', () => {
        expect(wrapper.find('Form').children()).toHaveLength(5);
        expect(wrapper.find("h1")).toHaveLength(1);
        expect(wrapper.find("h1").hasClass('page-title')).toEqual(true)
        const singupText = wrapper.find("h1").map(node => node.text());
        expect((singupText)).toEqual(["Sign Up"]);
        expect(wrapper.find("Input")).toHaveLength(4);
        expect(wrapper.find("Button")).toHaveLength(1);
    });

    it('+++ simulate change nome, sobrenome, email, password input', () => {
        const FirstName = {
            target: { name: 'nome', value: 'test' }
        };
        const SecondName = {
            target: { name: 'sobrenome', value: 'test' }
        };
        const eventEmail = {
            target: { name: 'email', value: 'test@test.com' }
        };
        const eventPassword = {
            target: { name: 'password', value: '123456' }
        };

        wrapper.find('[name="nome"]').simulate('change', FirstName);
        wrapper.find('[name="sobrenome"]').simulate('change', SecondName);
        wrapper.find('[name="email"]').simulate('change', eventEmail);
        wrapper.find('[name="password"]').simulate('change', eventPassword);

        expect(wrapper.state('nome')).toEqual({ "errorMsg": null, "validateStatus": "success", "value": "test" });

        expect(wrapper.state('sobrenome')).toEqual({ "errorMsg": null, "validateStatus": "success", "value": "test" });

        expect(wrapper.state('email')).toEqual({ "errorMsg": null, "validateStatus": null, "value": "test@test.com" });

        expect(wrapper.state('password')).toEqual({ "errorMsg": null, "validateStatus": "success", "value": "123456" });

        expect(wrapper.find('[name="nome"]').prop('value')).toEqual('test');

        expect(wrapper.find('[name="sobrenome"]').prop('value')).toEqual('test');

        expect(wrapper.find('[name="email"]').prop('value')).toEqual('test@test.com');

        expect(wrapper.find('[name="password"]').prop('value')).toEqual('123456');
    });

    it('+++ simulate fail in change nome, sobrenome, email, password input', () => {
        const FirstName = {
            target: { name: 'nome', value: 'tes' }
        };
        const SecondName = {
            target: { name: 'sobrenome', value: 'ts' }
        };
        const eventEmail = {
            target: { name: 'email', value: 'testetst.com.br' }
        };
        const eventPassword = {
            target: { name: 'password', value: '21' }
        };

        wrapper.find('[name="nome"]').simulate('change', FirstName);
        wrapper.find('[name="sobrenome"]').simulate('change', SecondName);
        wrapper.find('[name="email"]').simulate('change', eventEmail);
        wrapper.find('[name="password"]').simulate('change', eventPassword);

        expect(wrapper.state('nome')).toEqual({ "errorMsg": "Nome é muito curto (Mínimo 4 caractéres necessarios.)", "validateStatus": "error", "value": "tes" });

        expect(wrapper.state('sobrenome')).toEqual({ "errorMsg": "Sobrenome é muito curto (Mínimo 3 caractéres necessarios.)", "validateStatus": "error", "value": "ts" });

        expect(wrapper.state('email')).toEqual({ "errorMsg": "Email não é válido", "validateStatus": "error", "value": "testetst.com.br" });

        expect(wrapper.state('password')).toEqual({ "errorMsg": "Senha é muito curta (Mínimo 6 caractéres necessarios.)", "validateStatus": "error", "value": "21" });
    });

    it('+++ triggers submit handler with valid form data', () => {
        wrapper.find("Form").simulate('submit', { preventDefault: jest.fn() });
    });
});