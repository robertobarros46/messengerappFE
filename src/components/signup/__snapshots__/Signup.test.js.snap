// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup /> +++ render LoginPage correctly, snapshot 1`] = `
<div
  className="signup-container"
>
  <h1
    className="page-title"
  >
    Sign Up
  </h1>
  <div
    className="signup-content"
  >
    <Form
      className="signup-form"
      hideRequiredMark={false}
      layout="horizontal"
      onSubmit={[Function]}
      prefixCls="ant-form"
    >
      <FormItem
        colon={true}
        hasFeedback={false}
        label="Nome"
        prefixCls="ant-form"
      >
        <Input
          autoComplete="off"
          disabled={false}
          name="nome"
          onChange={[Function]}
          placeholder="Seu primeiro nome"
          prefixCls="ant-input"
          size="large"
          type="text"
          value=""
        />
      </FormItem>
      <FormItem
        colon={true}
        hasFeedback={false}
        label="Sobrenome"
        prefixCls="ant-form"
      >
        <Input
          autoComplete="off"
          disabled={false}
          name="sobrenome"
          onChange={[Function]}
          placeholder="Seu segundo nome"
          prefixCls="ant-input"
          size="large"
          type="text"
          value=""
        />
      </FormItem>
      <FormItem
        colon={true}
        hasFeedback={true}
        label="Email"
        prefixCls="ant-form"
      >
        <Input
          autoComplete="off"
          disabled={false}
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Seu email"
          prefixCls="ant-input"
          size="large"
          type="email"
          value=""
        />
      </FormItem>
      <FormItem
        colon={true}
        hasFeedback={false}
        label="Senha"
        prefixCls="ant-form"
      >
        <Input
          autoComplete="off"
          disabled={false}
          name="password"
          onChange={[Function]}
          placeholder="Um senha entre 6 e 20 caractéres"
          prefixCls="ant-input"
          size="large"
          type="password"
          value=""
        />
      </FormItem>
      <FormItem
        colon={true}
        hasFeedback={false}
        prefixCls="ant-form"
      >
        <Button
          block={false}
          className="signup-form-button"
          disabled={true}
          ghost={false}
          htmlType="submit"
          loading={false}
          prefixCls="ant-btn"
          size="large"
          type="primary"
        >
          Sign up
        </Button>
        Já é membro? 
        <Link
          replace={false}
          to="/login"
        >
          Faça login!
        </Link>
      </FormItem>
    </Form>
  </div>
</div>
`;
